core/src/crds_gossip_pull.rs

```diff
@@ -262,42 +267,71 @@ impl CrdsGossipPull {
                         if now > r.wallclock().checked_add(timeout).unwrap_or_else(|| 0)
                             || now + timeout < r.wallclock()
                         {
-                            timeout_count += 1;
-                            failed += 1;
+                            stats.timeout_count += 1;
+                            stats.failed += 1;
                             continue;
                         }
                     }
                     _ => {
                         // Before discarding this value, check if a ContactInfo for the owner
                         // exists in the table. If it doesn't, that implies that this value can be discarded
                         if crds.lookup(&CrdsValueLabel::ContactInfo(owner)).is_none() {
-                            timeout_count += 1;
-                            failed += 1;
+                            stats.timeout_count += 1;
+                            stats.failed += 1;
                             continue;
                         } else {
                             // Silently insert this old value without bumping record timestamps
-                            failed += crds.insert(r, now).is_err() as usize;
+                            match crds.would_insert(r, now) {
+                                Some(resp) => versioned_no_timestamp.push(resp),
+                                None => stats.failed += 1,
+                            }
                             continue;
                         }
                     }
                 }
             }
-            let old = crds.insert(r, now);
+            match crds.would_insert(r, now) {
+                Some(resp) => versioned.push(resp),
+                None => stats.failed += 1,
+            }
+        }
+        (versioned, versioned_no_timestamp)
+    }
+
+    /// process a vec of pull responses
+    pub fn process_pull_responses(
+        &mut self,
+        crds: &mut Crds,
+        from: &Pubkey,
+        responses: Vec<VersionedCrdsValue>,
+        responses_no_timeout: Vec<VersionedCrdsValue>,
+        now: u64,
+        stats: &mut ProcessPullStats,
+    ) {
+        let mut owners = HashSet::new();
+        for r in responses_no_timeout {
+            stats.failed += crds.insert_versioned(r).is_err() as usize;
```

Comment:
Do we want to distinguish between the different types of failures? (timeout vs duplicate)

Reply:
ok